src/ripple/basics/SlabAllocator.h

```diff
@@ -214,95 +268,111 @@ class SlabAllocator
         @return a pointer to a block of memory from the allocator, or
                 nullptr if the allocator can't satisfy this request.
      */
-    std::uint8_t*
+    [[nodiscard]] std::uint8_t*
     allocate() noexcept
     {
-        auto slab = slabs_.load();
+        auto active = active_.load();
 
-        while (slab != nullptr)
+        if (active)
         {
-            if (auto ret = slab->allocate())
+            if (auto ret = allocate(active))
                 return ret;
+        }
 
-            slab = slab->next_;
+        for (auto slab = slabs_.load(); slab; slab = slab->next_)
+        {
+            if (auto ret = allocate(slab))
+            {
+                // This may fail, but that's OK.
+                active_.compare_exchange_strong(active, slab);
+                return ret;
+            }
         }
 
         // No slab can satisfy our request, so we attempt to allocate a new
-        // one here:
-        std::size_t size = slabSize_;
-
-        // We want to allocate the memory at a 2 MiB boundary, to make it
-        // possible to use hugepage mappings on Linux:
+        // one. We align the block at a 2 MiB boundary to allow transparent
+        // hugepage support on Linux.
         auto buf =
-            boost::alignment::aligned_alloc(megabytes(std::size_t(2)), size);
+            reinterpret_cast<std::uint8_t*>(boost::alignment::aligned_alloc(
+                megabytes(std::size_t(2)), slabSize_));
 
-        // clang-format off
         if (!buf) [[unlikely]]
             return nullptr;
-            // clang-format on
 
 #if BOOST_OS_LINUX
-        // When allocating large blocks, attempt to leverage Linux's
-        // transparent hugepage support. It is unclear and difficult
-        // to accurately determine if doing this impacts performance
-        // enough to justify using platform-specific tricks.
-        if (size >= megabytes(std::size_t(4)))
-            madvise(buf, size, MADV_HUGEPAGE);
+        if (slabSize_ >= megabytes(std::size_t(4)))
+            madvise(buf, slabSize_, MADV_HUGEPAGE);
 #endif
 
-        // We need to carve out a bit of memory for the slab header
-        // and then align the rest appropriately:
-        auto slabData = reinterpret_cast<void*>(
-            reinterpret_cast<std::uint8_t*>(buf) + sizeof(SlabBlock));
-        auto slabSize = size - sizeof(SlabBlock);
+        // Check whether there's an existing slab with no associated buffer
+        // that we can give our newly allocated buffer to:
+        for (auto slab = slabs_.load(std::memory_order_relaxed);
```

Comment:
This loop does the load using `memory_order_relaxed` and the loop on line 282 uses `memory_order_seq_cst` (the default). Is this intentional? I _think_ they could both use the same memory order, and I _think_ they could both be relaxed (although I don't use memory orders that often, so I'm not confident here). If there's a reason for the difference we should probably add a comment. If there's no reason for the difference we should be consistent.

Reply:
@nbougalis  I'm leaving a comment here so we can have a threaded conversation, even though it doesn't relate to this particular line of code (github top comments can't be threaded).

I'm trying to think though the memory orders. What they do is define dependencies between the loads and stores of different atomic variables in the code. What I'm thinking about is the dependencies between the atomic variables `slabs_` and `active_`. For example, would the code blow up if a write to `active_` was seen before a write to `slabs_` and the active slab was not part of the slabs collection? What are the constraints here?

As far as I can tell, there is no dependency between `slabs_` and `active_` and the code works fine if `active_` is not part of the `slabs_` collection. I can't think of other constraints. If there aren't any other constraints, can't *all* the memory orders be `relaxed`? If there are constraints that I'm missing, can we document them so I can look at the code and make sure the memory orders will satisfy the constraints.

As for right now, I believe ALL the memory orders can be `relaxed`, but I may be missing important constraints.

Reply:
@scottschurr I agree! Using anything other than `std::memory_order::seq_cst`  is an optimization (often times it's a micro- or pico-optimization) and can be tricky and error-prone, even for experts, and I certainly don't consider myself an expert at this.

Given that we share that belief, it is mildly ironic that I am trying to use anything other than `std::memory_order::seq_cst` (and I think I'm the only one that is trying to do this in the codebase).

With that said, I do believe that it's worth trying to optimize code, especially low-level and performance critical code, and I don't think we should shy away from tools (even "dangerous" power-tools) that allow us to do this, just as long as we are careful and disciplined in their use.

I am fairly confident the semantics here are correct, but I will carefully re-think them (which, I think, is what @seelabs was trying to do as well). I will also run some benchmarks to determine the performance impact of using `std::memory_order::seq_cst` as opposed to `std::memory_order::relaxed`; if the performance impact is not measurable, then the increased "safety" and "understandability" of `std::memory_order::seq_cst` might make sense even if `std::memory_order::relaxed` semantics are fine.

Reply:
@nbougalis, thanks for offering to run those benchmarks.  Yes, if the performance impact of the riskier `memory_order` is measurable, then the benefits may justify the risks.

Reply:
Just BTW, here's a quote from the book "Rust Atomics and Locks", a very good book on this topic, even for C++ people like us:
```
While it might seem like the easiest memory ordering to reason about, SeqCst ordering is almost never necessary in practice. In nearly all cases, regular acquire and release ordering suffice.
```

So yeah, `relaxed` is difficult to reason about. `acquire/release` is much less difficult. I'd be OK with `acquire/release` over `seq_cst` (especially since as far as I can tell `relaxed` would work too). But of course, all of this is moot if there's no perf gain.

I'm also not sure how `relaxed` memory ordering work on intel. I vaguely remember that on intel there's no difference between `relaxed` and `acquire/release`, although I may be misremembering. I'll search for this when I get a chance...

Reply:
> @nbougalis, thanks for offering to run those benchmarks. Yes, if the performance impact of the riskier `memory_order` is measurable, then the benefits may justify the risks.

So, I ran some benchmarks on _synthetic code_ that implements the same "linked list" behavior that we have here, comparing the performance of `std::memory_order::relaxed` against `std::memory_order::seq_cst`.

Depending on the amount of concurrency (i.e. how many threads compete to modify the linked list), it appears that `std::memory_order::seq_cst` is between 25% and 100% slower that `std::memory_order::relaxed`. Here's a sample run (compiled with `-O3`):

```
------------------------------------------------------
Benchmark            Time             CPU   Iterations
------------------------------------------------------
relaxedTest   91517753 ns       513274 ns          100
cstTest      192449311 ns       994084 ns          100
```

Disclaimer: I _can't_ be sure if the test fully isolates the relevant code and eliminates all other influences, but the results generally seem to  suggest that we are accurately capturing the overhead of `std::memory_order::seq_cst`.

I am confident that the code does what is intended, whether we use `std::memory_order::relaxed` or `std::memory_order::seq_cst`, and so believe that it's hard to justify (the admittedly small) additional overhead of the latter. But if the consensus is that using `std::memory_order::seq_cst` is the desired option, I'm fine with that.

Reply:
Another option is `acquire/release`; I suspect you won't see any perf difference between `relaxed` and `acquire/release` (at least on intel platforms). If that's true, that would probably be my vote. As per the rust book referenced above `seq_cst` is almost never needed, and we don't need it in this case. If I'm wrong and there _is_ a perf difference, then we can look at `relaxed`.